### PR TITLE
ETC-11: Add TypeChecker command

### PR DESCRIPTION
**Pull request by shonfeder**
Also adds type checking passes into parser

This adds the final components of #244

### CODE REVIEW

File: tla-import/src/test/scala/at/forsyte/apalache/tla/imp/TestSanyImporter.scala [hunk 1]
```diff
@@ -1499,6 +1499,27 @@ class TestSanyImporter extends FunSuite {
     assert(aOfM.isInstanceOf[TlaOperDecl] && aOfM.asInstanceOf[TlaOperDecl].formalParams.size == 1)
   }
 
+  test("substitution in INSTANCE") {
```
Comment (shonfeder) [inline tla-import/src/test/scala/at/forsyte/apalache/tla/imp/TestSanyImporter.scala:1502]: @konnov is this related to the Etc work? Is it because of a regression that was introduced in the etc stuff, or is it a misplaced change?

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/TypeCheckerTool.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import at.forsyte.apalache.tla.typecheck.etc._
+
+/**
+  * The API to the type checker. It first translates a TLA+ module into EtcExpr and then does the type checking.
+  *
+  * @author Igor Konnov
+  */
+class TypeCheckerTool {
+  def check(listener: TypeCheckerListener, module: TlaModule): Boolean = {
+    // TODO: remember to type check the assumptions!
```
Comment (shonfeder) [inline tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/TypeCheckerTool.scala:13]: @konnov Is this something we should implement as part of this PR, or an additional feature we want to add later?

Comment (konnov) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/TypeCheckerTool.scala:13]: This is an additional feature. We have not been paying attention to `ASSUME(...)` so far, but we have to do it at some point.

File: tla-import/src/test/scala/at/forsyte/apalache/tla/imp/TestSanyImporter.scala (hunk 1 above)
Comment (konnov) [reply tla-import/src/test/scala/at/forsyte/apalache/tla/imp/TestSanyImporter.scala:1502]: I think it was just a quick fix/test for a type-checking benchmark that caused a failure in the parser.

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/etc/TestTypeCheckerTool.scala [hunk 3]
```diff
@@ -0,0 +1,69 @@
+package at.forsyte.apalache.tla.typecheck.etc
```
Comment (shonfeder) [inline tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/etc/TestTypeCheckerTool.scala:1]: @konnov Seems like we should also add tests for the many Tla modules that are part of this PR. 

Do you think it makes sense to do those as integration tests via our mdx setup? Or should I put them here?

Comment (konnov) [reply tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/etc/TestTypeCheckerTool.scala:1]: The PR has several `*.tla` files that can be used for integration testing. Yes, it would be good to have them integrated in the mdx setup.

Comment (shonfeder) [reply tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/etc/TestTypeCheckerTool.scala:1]: So currently type checking is failing on these files:

Comment (konnov) [reply tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/etc/TestTypeCheckerTool.scala:1]: Yes, it was definitely failing on `SlidingPuzzlesTyped` and `QueensTyped`. `QueensTyped` definitely require another operator that turns a function into a sequence. I don't remember about `TwoPhaseTyped.tla`; that is a bit surprising, but something was probably tricky there. Let's not fix these ones.

**Review by konnov — APPROVED** [review]
Looks good to me!

File: test/tla/SlidingPuzzlesTyped.tla [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+(***************************************************************************)
+(* Given a position and a unit translation vector return a pair of         *)
+(* pieces, before and after translation in opposite this vector direction  *)
+(***************************************************************************)      
+move(p, d) == "(<<Int, Int>>, <<Int, Int>>) => <<Int, Int>>" ##
+              LET s == "<<Int, Int>>" ##
+                    <<p[1] + d[1], p[2] + d[2]>>
+                  pc == ChooseOne(board, LAMBDA pc : s \in pc)
+              IN <<pc, {<<q[1] - d[1], q[2] - d[2]>> : q \in pc}>>
```
[the remark this answers is not in the document]
Comment (konnov) [reply test/tla/SlidingPuzzlesTyped.tla:49]: Are you talking about the type of `move(p, d)` or the type of `s`?

[the remark this answers is not in the document]
Comment (shonfeder) [reply test/tla/SlidingPuzzlesTyped.tla:49]: The type of `move`. 

It's annotated as `"(<<Int, Int>>, <<Int, Int>>) => <<Int, Int>>"` but the `LET` binding evaluates to 

```
<<pc, {<<q[1] - d[1], q[2] - d[2]>> : q \in pc}>>
```

`pc` is the result of apply `ChooseOne`, whose return type is `Set(<<Int, Int>>)`, and the second element of the tuple is also evidently a `Set(<<Int, Int>>)`. So I'm not sure where the `<<Int, Int>>` in the return type of `move`s annotation is supposed to come in. But also, from the way `move` is actually used, it looks like it should yield an  `<<Int, Int>>`. So that makes me wonder if I'm just misreading the spec in some basic way?

[the remark this answers is not in the document]
Comment (shonfeder) [reply test/tla/SlidingPuzzlesTyped.tla]: lol. I was misreading the usage of `move`. The type annotation was just mistaken :) I corrected it and things are working as expected now. :+1:

File: tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/TypeCheckerTool.scala (hunk 2 above)
Comment (shonfeder) [reply tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/TypeCheckerTool.scala:13]: Opened an issue to track this #330

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/etc/TestTypeCheckerTool.scala (hunk 3 above)
Comment (shonfeder) [reply tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/etc/TestTypeCheckerTool.scala:1]: I fixed the errors on `SlidingPuzzles` and `TwoPhased` and added a failing test for the `Queens` spec, which we can fix once support for the needed sub-typing or coercion is supplied.
